### PR TITLE
Fix array generation and recycling

### PR DESCRIPTION
**Pull request by PerfectLaugh**
Previous HeapAlloc2dArray failed under HeapPop due to incorrect heap layout.

### CODE REVIEW

File: vm/plugin-context.cpp [hunk 1]
```diff
@@ -497,7 +497,6 @@ PluginContext::popTrackerAndSetHeap()
 int
 PluginContext::pushTracker(uint32_t amount)
 {
-  assert(usesHeapTracker());
```
Comment (dvander) [inline vm/plugin-context.cpp:500]: The tracker must not be used in >= 1.11 binaries. Can you clarify what is being fixed here? I didn't see an explanation or test case.

File: vm/shell.cpp [hunk 2]
```diff
@@ -233,15 +233,17 @@ static cell_t Copy2dArrayToCallback(IPluginContext* cx, const cell_t* params)
 
   cell_t addr;
   if (!cx->HeapAlloc2dArray(params[2], params[3], &addr, flat_array))
-    return 0;
+    return cx->ThrowNativeError("Could not allocate %d * %d 2D Array: %d", params[2], params[3], err);
```
Comment (dvander) [inline vm/shell.cpp:236]: Functions returning "err" have already called ReportError, and don't have an error code.

File: vm/plugin-context.cpp (hunk 1 above)
Comment (PerfectLaugh) [reply vm/plugin-context.cpp:500]: The problem is HeapPop doesn't deallocate the resource allocated by HeapAlloc2dArray

https://github.com/alliedmodders/sourcemod/blob/fix-menu-vote-handler/core/logic/smn_menus.cpp#L510

Returns error 5. SP_ERROR_INVALID_ADDRESS.

Comment (PerfectLaugh) [reply vm/plugin-context.cpp:500]: image
This is the error from mapchooser_extended after running a while

File: vm/shell.cpp (hunk 2 above)
Comment (PerfectLaugh) [reply vm/shell.cpp:236]: Sorry I forgot to remove the left messages

File: vm/plugin-context.cpp (hunk 1 above)
Comment (PerfectLaugh) [reply vm/plugin-context.cpp:500]: I didn't know the tracker usage.

Comment (dvander) [reply vm/plugin-context.cpp:500]: Okay, thanks. This API does feel very weird to me. I'll add an RAII thing (eg "AutoHeapScope") to make this way easier to use. The current API is too difficult.

**Pull request closed without merging**